src/ui_imgui.hh

```diff
@@ -2654,6 +2655,68 @@ namespace UiLevelInfo {
     }
 }
 
+namespace UiConfirm {
+    static bool do_open = false;
+    std::string confirm_text = "";
+    std::string confirm_button1 = "";
+    std::string confirm_button2 = "";
+    std::string confirm_button3 = "";
+    int confirm_action1 = 0;
+    int confirm_action2 = 0;
+    int confirm_action3 = 0;
+    void *confirm_action1_data = 0;
+    void *confirm_action2_data = 0;
+    void *confirm_action3_data = 0;
+
+
+    void open(const char* text, const char* button1, principia_action action1,
+              const char* button2, principia_action action2, const char* button3, principia_action action3) {
+        confirm_text = text;
+
+        confirm_button1 = button1;
+        confirm_button2 = button2;
+        confirm_button3 = button3;
+
+        confirm_action1 = action1.action_id;
+        confirm_action2 = action2.action_id;
+        confirm_action3 = action3.action_id;
+
+        confirm_action1_data = action1.action_data;
+        confirm_action2_data = action2.action_data;
+        confirm_action3_data = action3.action_data;
+
+        do_open = true;
+    }
+
+    void layout() {
+        if (do_open) {
```

Comment:
I don't think there's any need to wrap this in an `if (do_open)`. `handle_do_open` handles it by itself and the rest are no-op if the dialog isn't open.